lib/profiling/nvidia/nvtx_prof.f90

```diff
@@ -0,0 +1,181 @@
+       use iso_c_binding
+       ! Have to use length of 1 because routine is BIND(C)
+       character(kind=C_CHAR, len=1) :: name
+     end subroutine nvtxRangePushA
+
+     ! Push range with custom label and custom colour
+     subroutine nvtxRangePushEx(event) bind(C, name='nvtxRangePushEx')
+       use iso_c_binding
+       import :: nvtxEventAttributes
+       type(nvtxEventAttributes) :: event
+     end subroutine nvtxRangePushEx
+  end interface nvtxRangePush
+
+  interface nvtxRangePop
+     subroutine nvtxRangePop() bind(C, name='nvtxRangePop')
+     end subroutine nvtxRangePop
+  end interface nvtxRangePop
+
+  ! Only the routines making up the PSyclone profiling API are public
+  public ProfileInit, ProfileFinalise, ProfileStart, ProfileEnd
+
+contains
+
+  !> An optional initialisation subroutine. This is not used for the NVTX
+  !! library.
+  subroutine ProfileInit()
+    implicit none
+    return
+  end subroutine ProfileInit
+
+  !> Starts a profiling area. The module and region name can be used to create
+  !! a unique name for each region.
+  !! Parameters: 
+  !! module_name:  Name of the module in which the region is
+  !! region_name:  Name of the region (could be name of an invoke, or
+  !!               subroutine name).
+  !! profile_data: Persistent data that can be used by the profiling library
+  !!               but isn't in this case.
+  subroutine ProfileStart(module_name, region_name, profile_data)
+    implicit none
+    character*(*), intent(in) :: module_name, region_name
+    type(ProfileData) :: profile_data
+
+    ! Round-robin the colour of each region created. Although nvtxStartRange
+    ! does a mod() on the id, we don't want to risk overflow so we don't
+    ! just blindly increment it.
+    if (last_colour < NUM_COLOURS) then
+       last_colour = last_colour + 1
+    else
+       last_colour = 1
+    end if
+    call nvtxStartRange(trim(module_name)//":"//region_name, &
+                        id=last_colour)
+    
+  end subroutine ProfileStart
+
+  !> Ends a profiling area.
+  !! profile_data: Persistent data, not used in this case.
+  subroutine ProfileEnd(profile_data)
+    implicit none
+    type(ProfileData) :: profile_data
+    
+    call nvtxEndRange()
+    
+  end subroutine ProfileEnd
+
+  !> The finalise function would normally print the results. However, this
+  !> is unnecessary for the NVTX library.
+  subroutine ProfileFinalise()
+    implicit none
+    return
+  end subroutine ProfileFinalise
+
+  !> Wrapper routine to create an NVTX Range. If present, id is an index
+  !! into the `col` array of colours.
+  subroutine nvtxStartRange(name, id)
+    implicit none
+    !> The name of the region we are starting
+    character(kind=C_CHAR, len=*), intent(in) :: name
+    !> Which colour to use for this region (indexes into the `col` array)
+    integer,             optional, intent(in) :: id
+    ! Locals
+    type(nvtxEventAttributes) :: event
+
+    tempName = trim(name)//C_NULL_CHAR
```

Comment:
I assume that nvtxRangePush keeps a copy of the name? Even so, adding \0 here adds another memcopy, which could be done only once.

Reply:
Removed.